FIX: Empty Links shows on frontend

tg-ephox opened the pull request:
Fixes https://github.com/WordPress/gutenberg/issues/2080

Changed the Toolbar to only create a link when the URL has been entered, instead of creating the link up front and then setting the URL.

I needed to lift the state for the link functionality from `FormatToolbar` to `Editable` as this needs to be reset from `onNodeChange`.  This allowed `FormatToolbar` to be simplified a lot as we don't have to guess what has happened in `componentWillReceiveProps`.

A remaining issue is that when typing into `UrlInput`, the cursor position is reset to the end if typing in the middle.  This is caused by the `updateSuggestions` function which does an api call and sets the state causing a re-render.  I believe that moving this to a separate component will fix the issue. (Using the separate component's render cycle instead of re-rendering the input which is confusing react as to where the cursor should be)

Fixes https://github.com/WordPress/gutenberg/issues/1069 (I put this in the same PR as its right in the same area)

Added CSS class and custom selector to TinyMCE so when the Toolbar is used, it doesn't loose focus.  Also before applying formatting, focus the TinyMCE instance.  This meant that the bookmark was no longer needed in state.

tg-ephox committed:
Lift add/edit link state to from FormatToolbar to Editable

tg-ephox committed:
Remove unused code

tg-ephox committed:
Remove unused code

tg-ephox committed:
Fix FireFox add link error

youknowriad commented on blocks/editable/format-toolbar/index.js:
```diff
@@ -42,55 +37,32 @@ const DEFAULT_CONTROLS = [ 'bold', 'italic', 'strikethrough', 'link' ];
-			linkValue: props.formats.link ? props.formats.link.value : '',
-			isEditingLink: false,
-		};
+		super( props );
```
I guess we should always prefer `super( ...arguments )` to avoid errors when we use the `context` feature?

tg-ephox replied:
No problems, i will revert this.

youknowriad commented on blocks/editable/index.js:
```diff
@@ -68,18 +68,26 @@ export default class Editable extends Component {
+				isAddingLink: false,
+				isEditingLink: false,
+				newLinkValue: '',
+			},
```
Why can't we keep this state in the `formatToolbar` instead. Seems more related to the link modal than the Editable.

tg-ephox replied:
These all need to be reset in `onNodeChange` which `FormatTooblar` has no knowledge of.  This also removes the need for trying to reconcile state in `componentWillReceiveProps` and the `setTimeout` hack that was in `addLink`.  I can't think of a cleaner way to do this but happy to take suggestions!

youknowriad replied:
I have a preference over `componentWillReceiveProps`. It's a better Separation of Concerns IMO but I can live with it. Thoughts @aduth?

aduth replied:
First reaction before digging in: I don't see it as a win to drop some lines and logic from `FormatToolbar` if it means we add more code and responsibilities to the already-bloated Editable.

aduth replied:
I'm not super familiar with this code, but it seems like we're trying to force this component to be treated as a [controlled component](https://facebook.github.io/react/docs/forms.html#controlled-components), maintaining the value as it's changed? This seems like a good use case for an [uncontrolled component](https://facebook.github.io/react/docs/uncontrolled-components.html), meaning the link toolbar starts with an empty initial value when first shown, and only at the point that we decide we want to confirm a link being added do we trigger the `onAddLink` callback.

youknowriad reviewed (commented):
This works great, but I think we should still keep the local state in the `FormatToolbar`. A state that is not used in a component but just passed as a prop to the child component should leave in the child component, don't you think?

tg-ephox committed:
Pass arguments to super

aduth commented:
>Added CSS class and custom selector to TinyMCE so when the Toolbar is used, it doesn't loose focus.

Can you explain why this is necessary? I'm a bit reluctant about this change for a few reasons:

- The Editable becomes aware of and dependent upon the existence of the Editor wrapper assigning this class
- So far as distinct Editables, and blocks not leveraging Editable, it would seem to me that we would _want_ blur to occur when leaving any individual Editable

tg-ephox committed:
Moved FormatToolbar state from Editable back to FormatToolbar

tg-ephox commented:
@aduth @youknowriad I removed the CSS class and selector, I was recommended to try this in order that the editor keeps its selection but it seems that focussing the editor before  applying/removing styling fixes this anyway.  Have also moved the `FormatToolbar` state back and added a `selectedNodeId` so that `FormatToolbar` can track node changes.

tg-ephox committed:
Remove unused code

tg-ephox committed:
Merge branch 'master' into fix/2080-empty-link-shows

tg-ephox committed:
Fix onKeyDown

youknowriad reviewed (approved):
It looks like this is working great. 👍 

When I worked on the UrlInput the first time, I had to do add this `convert_urls: false,` to the TinyMCE settings to avoid JS errors when using empty URLs. Maybe we can remove this option now?

tg-ephox commented:
@youknowriad I tried removing the `convert_urls` setting from TinyMCE and didn't see any difference... Although `UrlInput` forces a complete URL to be entered anyway so this option might not be relevant, but I have left it there for now.
